src/data/nat/factorial.lean

```diff
@@ -123,75 +130,216 @@ begin
+theorem factorial_mul_asc_fact (n : ℕ) : ∀ k, n! * n.asc_fact k = (n + k)!
+| 0 := by rw [asc_fact, add_zero, mul_one]
+| (k + 1) := by rw [asc_fact_succ, mul_left_comm, factorial_mul_asc_fact, ← add_assoc, factorial]
 
-/-- Avoid in favour of `nat.factorial_mul_desc_fac` if you can. ℕ-division isn't worth it. -/
-lemma desc_fac_eq_div (n k : ℕ) : desc_fac n k = (n + k)! / n! :=
+/-- Avoid in favour of `nat.factorial_mul_asc_fact` if you can. ℕ-division isn't worth it. -/
+lemma asc_fact_eq_div (n k : ℕ) : n.asc_fact k = (n + k)! / n! :=
 begin
   apply mul_left_cancel' (factorial_ne_zero n),
-  rw factorial_mul_desc_fac,
+  rw factorial_mul_asc_fact,
   exact (nat.mul_div_cancel' $ factorial_dvd_factorial $ le.intro rfl).symm
 end
 
-lemma desc_fac_of_sub {n k : ℕ} (h : n < k) :
-  (k - n) * (k - n).desc_fac n = (k - (n + 1)).desc_fac (n + 1) :=
+lemma asc_fact_of_sub {n k : ℕ} (h : k < n) :
+  (n - k) * (n - k).asc_fact k = (n - (k + 1)).asc_fact (k + 1) :=
 begin
-  set t := k - n.succ with ht,
-  suffices h' : k - n = t.succ, by rw [←ht, h', succ_desc_fac, desc_fac_succ],
+  set t := n - k.succ with ht,
+  suffices h' : n - k = t.succ, by rw [←ht, h', succ_asc_fact, asc_fact_succ],
   rw [ht, succ_eq_add_one, ←sub_sub_assoc h (succ_pos _), succ_sub_one],
 end
 
-end desc_fac
+lemma pow_le_asc_fact (n : ℕ) : ∀ (k : ℕ), (n + 1)^k ≤ n.asc_fact k
+| 0 := by rw [asc_fact_zero, pow_zero]
+| (k + 1) := begin
+  rw [asc_fact_succ, pow_succ],
+  exact nat.mul_le_mul (nat.add_le_add_right le_self_add _) (pow_le_asc_fact k),
+end
+
+lemma pow_lt_asc_fact (n : ℕ) : ∀ {k : ℕ}, 2 ≤ k → (n + 1)^k < n.asc_fact k
+| 0 := by rintro ⟨⟩
+| 1 := by rintro (_ | ⟨_, ⟨⟩⟩)
+| (k + 2) := λ _, begin
+  rw [asc_fact_succ, pow_succ],
+  exact nat.mul_lt_mul (nat.add_lt_add_right (nat.lt_add_of_pos_right (succ_pos k)) 1)
+    (pow_le_asc_fact n _) (pow_pos (succ_pos n) _),
+end
+
+lemma asc_fact_le_pow (n : ℕ) : ∀ (k : ℕ), n.asc_fact k ≤ (n + k)^k
+| 0 := by rw [asc_fact_zero, pow_zero]
+| (k + 1) := begin
+  rw [asc_fact_succ, pow_succ],
+  exact nat.mul_le_mul_of_nonneg_left ((asc_fact_le_pow k).trans (nat.pow_le_pow_of_le_left
+  (le_succ _) _)),
+end
+
+lemma asc_fact_lt_pow (n : ℕ) : ∀ {k : ℕ}, 2 ≤ k → n.asc_fact k < (n + k)^k
+| 0 := by rintro ⟨⟩
+| 1 := by rintro (_ | ⟨_, ⟨⟩⟩)
+| (k + 2) := λ _, begin
+  rw [asc_fact_succ, pow_succ],
+  refine nat.mul_lt_mul' (le_refl _) (lt_of_le_of_lt (asc_fact_le_pow n _)
+    (pow_lt_pow_of_lt_left (lt_add_one _) (succ_pos _))) (succ_pos _),
+end
+
+lemma asc_fact_pos (n k : ℕ) : 0 < n.asc_fact k :=
+lt_of_lt_of_le (pow_pos (succ_pos n) k) (pow_le_asc_fact n k)
+
+end asc_fact
+
+section desc_fact
+
+/-- n.desc_fact k = n! / (n - k)! (as seen in `nat.desc_fact_eq_div`), but implemented recursively
+to allow for "quick" computation when using `norm_num`. This is closely related to `pochhammer`, but
+much less general. -/
+def desc_fact : ℕ → ℕ → ℕ
```

Comment:
we should standardise how we pattern-match: i prefer the method I did before, as it makes it explicit that `n` is fixed